### ISSUE TITLE
Change MariaDB charset from utf8 to utf8mb4

### ISSUE THREAD

**Issue by hoyes** [camdram/camdram#448]
https://medium.com/@adamhooper/in-mysql-never-use-utf8-use-utf8mb4-11761243e434

Required in order to import Tweets that contain emojis after changes in #443

**Comment by hoyes** [comment]
DB migration added.

I hope this was a good idea... https://development.camdram.net/shows/2018- :smiling_imp: 

I kinda wonder if we should reopen #302...

**Comment by hoyes** [comment]
Reopening as something's not quite right (see Tweet on 7th July at https://www.camdram.net/societies/cambridge-university-amateur-dramatic-club). Works on my machine so looks like a config/version thing.

**Comment by CHTJonas** [comment]
Hmm weird that it seems to support the red heart emoji but not any of the others.

**Comment by philosophicles** [comment]
> I kinda wonder if we should reopen #302...

I would advocate new issue rather than reopen that mess, but I kinda sorta do feel it might be worth considering, actually? Given we live in a world where The Emoji Movie is a thing. Emojis could be a legitimate part of a new show's title. Just me?

**Comment by GKFX** [comment]
@CHTJonas Red heart wasn’t originally an emoji so it’s in the wrong block (U+2665 not U+1Fxxx).

**Comment by GKFX** [comment]
I don’t like the ides that `2018-😇` might be a slug, because on a many (most?) desktops it will a) look a bit off and b) be impossible to type. Instead, let’s say that if `slug` is invalid, the canonical URL of a show is `/shows/by-id/7395` or whatever. It’s not pretty but it would work, and if the show is later renamed to have letters in, it works as a redirect.

**Comment by CHTJonas** [comment]
Yes I agree - having emojis in the URL bar probably isn't the way to go. I can actually do it using Safari on macOS eg. https://github.com/search?q=😎 (and it shows up in the bar as a rendered emoji) but I get the feeling that here be dragons...

What we're saying is that a show/venue/society/advert/role/person can have an Emoji(s) in its name, but not in its slug. So we need a way of reducing the emoji to some form of text. I'm not sure I entirely like having by-id as the canonical URL but I also see no easy way to reduce a slug like `2018-😇` to something entirely sensible. We could maybe use unicode descriptions such as `2018-smiling-face-with-halo` but I imagine that something like `2018-the-emoji-show-😎💩❤️🙅🏻‍♀️` would generate a stupidly long slug off `2018-the-emoji-show-smiling-face-with-sunglasses-pile-of-poop-red-heart-woman-gesturing-no`

**Comment by GKFX** [comment]
I think that if a show is called `The Emoji Show 😎💩❤️🙅🏻‍♀️` then its slug should just be `2018-the-emoji-show`, i.e. we scrub the emojis. If the show is just called 😈 then we’re in a more awkward situation as by that method, the slug becomes `2018-`.

When that occurs we’d need a fallback method. Maybe chose the hex code of the emojis? `2018-1f4a2` or whatever looks bad but not that awful. I don’t think we can be expected to achieve a wonderful outcome here.

The alternative to all this is that we let producers choose their own slug, so that they can specify something that is meaningful and concise themselves.

**Comment by hoyes** [comment]
Hashing sounds reasonable. Customizable slugs is probably overkill.

There's already a custom slug generator for shows to handle inserting the year, so could maybe shoehorn this in there.

**Comment by CHTJonas** [comment]
This should maybe throw up some kind of warning too to let users know that we're ditching their emojis from the show's slug, or that we have generated a semi-artificial one as there was no text to use.

**Comment by GKFX** [comment]
I think @philosophicles has previously suggested that we make the URL appear under the show name field when the user is creating/editing a show, which is better than putting up a big warning.

**Comment by hoyes** [comment]
The original encoding issue is now fixed (https://www.camdram.net/societies/cambridge-university-amateur-dramatic-club, https://www.camdram.net/societies/camdram). An old .cnf files was still lurking from prior to the MariaDB upgrade which was forcing the encoding back to utf8 sans mb4 - now deleted.

**Comment by CHTJonas** [comment]
**Emoji-Support Task List:**
- [x] Change database encoding.
- [ ] Check that users and people (maybe venues too?) can't be created with emojis in the name.
- [x] Allow shows to be created with emojis in the name.
- [ ] Ignore emojis when generating show slugs if there is other text to use, hex-encode if there isn't.

Feel free to check off/add to the above 😄

**Comment by CHTJonas** [comment]
Just been doing some research. Seems there is already a way to encode non-ASCII in the URL in the form of [RFC 3492](https://tools.ietf.org/html/rfc3492) aka [Punnycode](https://en.wikipedia.org/wiki/Punycode). There's a converter [here](https://www.charset.org/punycode) if you want to play around and see how this works.

I was looking at this mainly in reference to the suggestion in #302 that there might be, for example, a Chinese actor or a Chinese-translated play. Just had a fiddle around on dev and what do you know: https://development.camdram.net/shows/2018-juan-tong-zhi (although this obviously isn't punnycode so I assume Symfony is doing some clever framework stuff in the background).

**Comment by hoyes** [comment]
That's kinda cool. Looks like it's using https://github.com/Behat/Transliterator, which is itself a PHP port of https://metacpan.org/pod/release/SBURKE/Text-Unidecode-0.04/lib/Text/Unidecode.pm, which is self-admittedly a slightly hacky and non-comprehensive approach to transliteration though. From my limited knowledge of Chinese, certain characters have different Pinyin forms in different contexts so if someone did put a Chinese-language show on Camdram then they might still complain that the slug contains an incorrect "translation" of the show title...

I'm wondering now whether we should not bother trying and put non-ASCII characters in the slug verbatim using [IRI encoding](https://en.wikipedia.org/wiki/Internationalized_Resource_Identifier) (it looks like Punnycode is more to do with host names...?). On in other words, if someone wants to include untypable characters in their show title on a site targeting an English-speaking audience then they're welcome and we'll make a matching slug, but don't come complaining thereafter if people have problems searching for it. Search now handles "asciification" of accented characters common in Western languages, which I'm guessing is how most people access show pages.

Wikipedia has https://en.wikipedia.org/wiki/Les_Mis%C3%A9rables (although it admittedly has a redirect from the non-accented version), and the Chinese-language Wikipedia has https://zh.wikipedia.org/wiki/%E6%82%B2%E6%83%A8%E4%B8%96%E7%95%8C, both of which are rendered more conventionally in the URL bar, and can still be copy+pasted.

**Comment by GKFX** [comment]
Re dubious Pinyin: it seems really unlikely that more than a few people will list chinese characters in their show on Camdram. If that then comes out wrong in the slug we can do a DB edit for them; I’d rather not over-automate something which might never happen.

**Comment by hoyes** [comment]
Yea true. The point was more that just allowing unicode characters in slugs IRI-style is a way of ensuring that each character always produces >=1 ASCII chars that have some non-controversial relation to the original for all possible unicode codepoints, and the fact that it looks a bit ugly in contexts that don't denormalize the %-encoded chars probably doesn't matter that much given this is an edge case.

**Comment by CHTJonas** [comment]
I'm just wondering - if we're changing/hooking into the transliterator to hash emojis, would it be all that more difficult to hash all non-ASCII as well? Either way I think it's more important that we display the slug to the user during the add show process as has been suggested before, and possibly with a note that they should contact websupport if they are unhappy with the rendering

**Updated Emoji-Support Task List:**
- [x] Change database encoding.
- [ ] Check that users and people (maybe venues too?) can't be created with emojis in the name.
- [x] Allow shows to be created with emojis in the name.
- [ ] Ignore emojis when generating show slugs if there is other text to use, hex-encode if there isn't.
- [ ] Display the slug to the user before they add the show so they can confirm they're happy

**Comment by GKFX** [comment]
I've edited one of the task list items above. Phrasing it like that means that we're not hooking into the transliterator in such a complex way as going in and hex-encoding the emojis, we just do a check for an empty string at some point. I'd suggest that just taking a hash of the whole string and truncating it might make more sense than hex-encoding each emoji. If anyone who isn't familiar with Unicode sees `🏴󠁧󠁢󠁥󠁮󠁧󠁿` (England flag) become `1f3f4-e0067-e0062-e0065-e006e-e0067-e007F` they are going to be mighty confused...
